### PR TITLE
find-esp: gracefully handle btrfs RAID

### PR DESCRIPTION
**Pull request by yuwata**
Fixes RHBZ#2251262 (https://bugzilla.redhat.com/show_bug.cgi?id=2251262).

### CODE REVIEW

**Comment by arcivanov** [comment]
Thank you!

**Comment by yuwata** [comment]
I have not tested this yet. I will try to add a short test later.

**Comment by arcivanov** [comment]
> @arcivanov Maybe, /boot is on btrfs RAID?
> 

Yes.

**Comment by arcivanov** [comment]
> @arcivanov Maybe, /boot is on btrfs RAID?

Also `/boot/efi` on `/dev/md0`, if it makes a difference.

**Comment by yuwata** [comment]
> > @arcivanov Maybe, /boot is on btrfs RAID?
> 
> Also `/boot/efi` on `/dev/md0`, if it makes a difference.

Thanks. I added testcases for md and btrfs RAID, and should be fine now.

File: src/shared/find-esp.c [hunk 1]
```diff
@@ -308,8 +308,23 @@ static int verify_fsroot_dir(
...
+
+                /* Hmm, maybe a btrfs device, and the caller asked for the backing device?
+                 * Then let's try to get it. */
+
+                r = btrfs_get_block_device_at(dir_fd, strempty(f), &devnum);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        return log_full_errno(searching ? LOG_DEBUG : LOG_ERR,
+                                              SYNTHETIC_ERRNO(searching ? EADDRNOTAVAIL : ENOTSUP),
```
Comment (poettering) [inline src/shared/find-esp.c]: use EOPNOTSUPP, which is how Linux primarily calls that error, not ENOTSUP

File: src/shared/find-esp.c [hunk 2]
```diff
@@ -308,8 +308,23 @@ static int verify_fsroot_dir(
...
+                                              "Directory \"%s\" is on a btrfs RAID, not supported.", path);
+
+                *ret_dev = devnum;
+                return 0;
+        }
```
Comment (poettering) [inline src/shared/find-esp.c]: hmm, shouldn't rather the callers of this function be fixed to handle a zero device num properly as a way of "we don't know the backing device".

i mean verify_fsroot_dir() is not primarily about returning the device node, but primarily abot detecting if we are at the top of the fs, and hence i think w shouldn't fail if we manage to detect that but just don't manage to detect that secondary bit of info, i.e. the backing devnum...

File: src/shared/find-esp.c [hunk 3]
```diff
@@ -396,6 +396,11 @@ static int verify_esp(
         if (relax_checks)
                 goto finish;
 
+        if (devnum_is_zero(devid))
+                return log_full_errno(searching ? LOG_DEBUG : LOG_ERR,
+                                      SYNTHETIC_ERRNO(searching ? EADDRNOTAVAIL : EOPNOTSUPP),
+                                      "Directory \"%s\" is on a btrfs RAID, not supported.", p);
```
Comment (poettering) [inline src/shared/find-esp.c]: I'd probably have said "Could not determine backing block device of directory %s (btrfs RAID?)" or so

File: src/shared/find-esp.c [hunk 4]
```diff
@@ -767,6 +772,14 @@ static int verify_xbootldr(
         if (relax_checks)
                 goto finish;
 
+        if (devnum_is_zero(devid))
+                return log_full_errno(searching ? LOG_DEBUG : LOG_ERR,
+                                      SYNTHETIC_ERRNO(searching ? EADDRNOTAVAIL : EOPNOTSUPP),
+                                      "Directory \"%s\" is on a btrfs RAID, not supported.\n"
```
Comment (poettering) [inline src/shared/find-esp.c]: and similar here

File: src/shared/find-esp.c [hunk 5]
```diff
@@ -767,6 +772,15 @@ static int verify_xbootldr(
         if (relax_checks)
                 goto finish;
 
+        if (devnum_is_zero(devid))
+                return log_full_errno(searching ? LOG_DEBUG : LOG_ERR,
+                                      SYNTHETIC_ERRNO(searching ? EADDRNOTAVAIL : ENODEV),
+                                      "Could not determine backing block device of directory \"%s\" (btrfs RAID?).%s",
+                                      p,
+                                      searching ? "" :
+                                      "Hint: set $SYSTEMD_RELAX_XBOOTLDR_CHECKS=yes environment variable "
```
Comment (keszybz) [inline src/shared/find-esp.c]: Please prefix with "\n" here, otherwise the two strings run together.

**Review by keszybz — CHANGES_REQUESTED** [review]
Hmm, I don't think I like this approach at all. Let's say that we're indeed on a system with btrfs raid. With this patch, we will **knowingly** return a wrong answer, creating a mess where things are installed or searched for in the wrong location. This seems to be even worse than the earlier case where we returned an error.

But returning of the error causes a problem for existing systems which are set up in this way, making people justifiably unhappy.

If the user set up ESP or XBOOTLDR on a multi-partition device, this was their choice. We don't recommend this, and it certainly has various problems with firmware, etc, but that is not our problem. If we cannot verify the backing device, then we should just not do the verification. I think we should print a warning: "/boot uses a multi-device filesystem, this is not supported. Skipping verification of backing device."

File: src/shared/find-esp.c (hunk 5 above)
Comment (yuwata) [reply src/shared/find-esp.c]: Oh, I mistakenly dropped `\n` on rebase.

**Comment by arcivanov** [comment]
> > Hmm, I don't think I like this approach at all. Let's say that we're indeed on a system with btrfs raid. With this patch, we will **knowingly** return a wrong answer, creating a mess where things are installed or searched for in the wrong location. This seems to be even worse than the earlier case where we returned an error.
> we are using this to search for the ESP, and /boot is one candidate for it, but if its btrfs raid

So in this case:

```
/dev/md127: UUID="9ED6-2928" BLOCK_SIZE="512" TYPE="vfat"
/dev/sda1: UUID="baa45d33-ce53-99ea-4caa-3cbcf106d41a" UUID_SUB="6133255f-c8d3-a254-22bf-0eb05b935b32" LABEL="0" TYPE="linux_raid_member" PARTLABEL="EFI System Partition" PARTUUID="6b4716d4-2e3b-46d8-9733-809ab5571b27"
/dev/sdb1: UUID="baa45d33-ce53-99ea-4caa-3cbcf106d41a" UUID_SUB="8ac53325-e23a-5f49-5880-6f3e579a8c6d" LABEL="0" TYPE="linux_raid_member" PARTUUID="a63ea092-385c-0d4a-8c2b-f00cac5d4ce3"

/dev/sda2: UUID="c54fb68b-52fe-4521-a73c-b7e18f33430f" UUID_SUB="95c04b40-3165-4fb0-86d2-60876f8cef3f" BLOCK_SIZE="4096" TYPE="btrfs" PARTUUID="f9336765-f8ce-4e23-826b-b13bce619bf0"
/dev/sdb2: UUID="c54fb68b-52fe-4521-a73c-b7e18f33430f" UUID_SUB="12d23515-8e55-4f8a-abf2-00b13ddb4107" BLOCK_SIZE="4096" TYPE="btrfs" PARTUUID="71f7dfab-2904-364e-879a-d1d7848807c9"
```

Where `md127` is the ESP and `sda2/sdb2` is the `btrfs` RAID. 

`sda2/sdb2 -> /boot`
`md127 -> /boot/efi` - ESP

As far as I understand (and admittedly my understanding is very rudimentary) ESP can only be vfat with GUID  	C12A7328-F81F-11D2-BA4B-00A0C93EC93B. 

So is it necessary to read the partitions that don't have the type GUID above  or don't have an appropriate FS type?

**Comment by yuwata** [comment]
I added several more commits, and more test cases.

Note,
- now setups like used by @arcivanov, that is, ESP is a md RAID partition and /boot is a (md, btrfs, or any) RAID partition, kernel-install use /boot as BOOT_ROOT, as kernel-install falls back to /boot when both find_esp() and find_xbootldr() fail.
- Also, it should work well if ESP is a md RAID partition, but /boot is a _non_-RAID partition, as kernel-install uses xbootldr partition if found.
- However, if ESP is a regular vfat partition, but /boot is a (md, btrfs, or any) RAID partition, then kernel-install will install kernel and initrd to ESP by default. So, in such situation, people needs to set `$SYSTEMD_RELAX_XBOOTLDR_CHECKS=yes` to make kernel-install use xbootldr partition.

> If the user set up ESP or XBOOTLDR on a multi-partition device, this was their choice. We don't recommend this, and it certainly has various problems with firmware, etc, but that is not our problem. If we cannot verify the backing device, then we should just not do the verification. I think we should print a warning: "/boot uses a multi-device filesystem, this is not supported. Skipping verification of backing device."

I think skipping check is dangerous, as we may pick wrong directory...

> We don't recommend this, and it certainly has various problems with firmware, etc, but that is not our problem.

Right. But, if we silently accept RAID partitions without verification, then people think such setup is _supported_. And if something broken (especially in the future), people claims that it worked and was supported. I think we should request people to explicitly opt-out the checks via env vars.

**Comment by arcivanov** [comment]
> * However, if ESP is a regular vfat partition, but /boot is a (md, btrfs, or any) RAID partition, then kernel-install will install kernel and initrd to ESP by default. So, in such situation, people needs to set `$SYSTEMD_RELAX_XBOOTLDR_CHECKS=yes` to make kernel-install use xbootldr partition.

This was not the behavior of the previous `kernel-install`  (when it was a script) and it will effectively break all the systems since ESP is usually not large enough. It worked for literal years previously without ANY problems and mirror boot partition is essential for high(-er) reliability systems.

**Comment by yuwata** [comment]
> > * However, if ESP is a regular vfat partition, but /boot is a (md, btrfs, or any) RAID partition, then kernel-install will install kernel and initrd to ESP by default. So, in such situation, people needs to set `$SYSTEMD_RELAX_XBOOTLDR_CHECKS=yes` to make kernel-install use xbootldr partition.
> 
> This was not the behavior of the previous `kernel-install` (when it was a script) and it will effectively break all such systems since ESP is usually not large enough. It worked for literal years previously without ANY problems and mirror boot partition is essential for high(-er) reliability systems.

Right.

**Comment by arcivanov** [comment]
Any update on this? This bricked kernel updates on Fedora 39 for pretty much everybody with a btrfs mirror /boot.

**Comment by poettering** [comment]
I am not convinced we should pretend that ESP on RAID was a supported setup. That's not supportable, since the firmware does not support reading from RAID generally, and considers the ESP a writable fs. 

Hence, I mean, i am fine if people set some env var, that basically says "Hey I want to shoot myself in the foot", but otherwise we should not consider ESP-on-RAID a valid setup.

**Comment by arcivanov** [comment]
> I am not convinced we should pretend that ESP on RAID was a supported setup. That's not supportable, since the firmware does not support reading from RAID generally, and considers the ESP a writable fs.

In this scenario the ESP is not on RAID and never was. 

/boot/efi is ESP
/boot is NOT and is a BTRFS mirror.

**Comment by poettering** [comment]
Hmm, this suggests otherwise?

https://github.com/systemd/systemd/pull/30321#issuecomment-1845544863

What is it now? Am I reading that wrong? That suggests that there was RAID on ESP?

**Comment by arcivanov** [comment]
> Hmm, this suggests otherwise?
> 
> [#30321 (comment)](https://github.com/systemd/systemd/pull/30321#issuecomment-1845544863)
> 
> What is it now? Am I reading that wrong? That suggests that there was RAID on ESP?

The md0 in this configuration is RAID1 but because the metadata superblock 1.0 (or 0.9) is at the end of the device the partition appears as a regular partition for UEFI BIOS.

This mechanism existed for decades specifically because it works with both Legacy and UEFI due to mirrored partition appearing as regular ones to the firmware.

https://raid.wiki.kernel.org/index.php/A_guide_to_mdadm#Superblocks_and_Raid_versions
https://www.linuxquestions.org/questions/linux-newbie-8/how-to-mirror-the-boot-partition-on-existing-raid1-865222/

The issue here is not the ESP /boot/efi but /boot which is a BTRFS mirror. 
Other users have encountered this same issue without a transparent md0 on the ESP, but simply because their /boot was a BTRFS mirror: https://bugzilla.redhat.com/show_bug.cgi?id=2251262#c12

**Comment by poettering** [comment]
Nope, sorry. ESP is a writable resource from EFI mode, hence any setup like this means the system will likely come up dirty regularly, because firmware only will touch one copy, not the other.

Again, ESP on RAID is not supported, and we should not pretend this was safe, because it isn't.

Yes, in specific corner cases this might work, because you are lucky that nothin in your boot chain writes a sector, but it's not safe in the general case, and generally counterproductive as it doesn't increase resilience but decrease it.

Hence, I am fine with supporting this if people explicitly request something like this, but in automatic cases this is fishy and we should step away.

File: src/shared/find-esp.c [hunk 6]
```diff
@@ -27,9 +27,29 @@
...
+        VerifyESPFlags flags = 0;
+
+        assert(env_name_for_relaxing);
+
+        if (unprivileged_mode < 0)
+                unprivileged_mode = geteuid() != 0;
+        if (unprivileged_mode)
+                flags |= VERIFY_ESP_UNPRIVILEGED_MODE;
+
+        if (getenv_bool(env_name_for_relaxing) > 0)
```
Comment (poettering) [inline src/shared/find-esp.c:44]: maybe if you touch this also do:

```c
r = getenv_bool(…);
if (r < 0 && r != -ENXIO) 
         log_debug_errno(r, "Failed to parse $%s value, assuming false.", …);
```
you get the idea

**Comment by poettering** [comment]
PR looks good though

**Comment by arcivanov** [comment]
How do we initiate a backport?

**Comment by mrc0mmand** [comment]
> How do we initiate a backport?

Setting the `needs-stable-backport` label is enough - once we start collecting stuff for the next stable releases, this PR will be picked up as well.

**Comment by keszybz** [comment]
I queued this up for v255.3.

**Comment by arcivanov** [comment]
> I queued this up for v255.3.

Can this be backported to v253.16?

**Comment by arcivanov** [comment]
And v254.9?

**Pull request merged**